[Communication] - SDK - Adding Contributing Guide (#20876)
Azure/azure-sdk-for-java: sdk/communication/CONTRIBUTING.md

# Contributing Guide

> This is a contributing guide made specifically for the Azure Communication Services SDK. The Azure SDK repo also has a contributing guide that might help you in some other general processes. If you haven't checked that one out yet, you can find it [here](https://github.com/Azure/azure-sdk-for-java/blob/master/CONTRIBUTING.md). This guide assumes you have set up your development environment for this repository.

Reviewer: repo -> repository

The Azure Communication Services SDK for Java currently consists of 5 different packages. While each package has its own set of environment variables to make their tests run successfully, all of them follow a similar structure that allows a smooth onboarding process.

> Let's get started with how to setup the repository itself.

Reviewer: ```suggestion
Let's get started with additional setup steps for this package.
```

When you go inside the tests folder of the package you are working with, you will see a folder called `resources` that has another folder named `session-records`. This folder contains, as its name suggests, recordings of successful calls to the API that allow us to run the tests in PLAYBACK mode and remove the necessity of hitting the actual resources every time we may want to test.

### Playback mode

To run the tests in PLAYBACK mode, set an environment variable called `AZURE_TEST_MODE` and set its value to `PLAYBACK` (If the variable if not set, the default will be `PLAYBACK`). After your variable has been set, change directory to the root folder of the package you're working on and run the `mvn verify` command.

> If the tests are successful, we can proceed to run the tests in LIVE mode.

Reviewer: It is not required to run playback before running live mode, correct? If that is true, this sentence is not necessary.

### Live mode

Because in LIVE mode we are hitting an actual resource, we must set the appropriate environment variable to make sure the code tests against the resource we want. Set up two env variables called `COMMUNICATION_LIVETEST_CONNECTION_STRING` and `COMMUNICATION_CONNECTION_STRING` and set them to the connection string of the target communication resource you want to test against. 

> 

Reviewer: The value of the connection strings can be obtained from Azure portal. [Access your connection strings and service endpoints ] (https://docs.microsoft.com/en-us/azure/communication-services/quickstarts/create-communication-resource?tabs=windows&pivots=platform-azp)

You can run the `mvn verify` command after setting the `AZURE_TEST_MODE` variable to `LIVE`.

### Record mode

RECORD mode is similar to LIVE mode because it also hits an actual resource. In addition to hitting the resource, RECORD mode will also record the successful calls to the service in json format. As mentioned, the recordings are stored in the package root folder under `target/test-classes/session-records`.

These newly generated files will have to be copied to the previosly mentioned `resources/session-records` after completion to make sure the PLAYBACK tests run with an updated version of the calls we made. Make sure to change the name of the recording files to match the names of the ones that are already in the `resources/session-records` folder. 

> If you would like to generate new recordings for a test, setting `AZURE_TEST_MODE` to `RECORD`, reopen Visual Studio, and run the tests normally.

Reviewer: ```suggestion
If you would like to generate new recordings for a single test, setting `AZURE_TEST_MODE` to `RECORD`, reopen Visual Studio, and run the test in Visual Studio normally.
```

## Testing

Make sure to check out the general contributing guide the Azure SDK repo has for a more in-depth look at testing and setting up your dev environment. You can check out the contributing file [here](https://github.com/Azure/azure-sdk-for-java/blob/master/CONTRIBUTING.md)

When you go inside the tests folder of the package you are working with, you will see a folder called `resources` that has another folder named `session-records`. This folder contains, as its name suggests, recordings of successful calls to the API that allow us to run the tests in PLAYBACK mode and remove the necessity of hitting the actual resources every time we may want to test.

### Playback mode

To run the tests in PLAYBACK mode, set an environment variable called `AZURE_TEST_MODE` and set its value to `PLAYBACK` (If the variable if not set, the default will be `PLAYBACK`). After your variable has been set, change directory to the root folder of the package you're working on and run the `mvn verify` command.

If the tests are successful, we can proceed to run the tests in LIVE mode.

### Live mode

> Because in LIVE mode we are hitting an actual resource, we must set the appropriate environment variable to make sure the code tests against the resource we want. Set up an env variable called `COMMUNICATION_LIVETEST_CONNECTION_STRING` and set it to the connection string of the resource you want to test against. 

Reviewer 2: `COMMUNICATION_LIVETEST_CONNECTION_STRING` is only used for SMS and Phone Numbers. We use `COMMUNICATION_CONNECTION_STRING` for the rest of the SDKs

> Depending on which package you are testing, it may need special environment variables to test succesfully. All packages have a TestBase file inside their corresponding test folder and each one of these contain the special environment variables the tests need in order to run. Make sure to set these variables before running the tests themselves. You may need to restart your development environment after creating or updating these environment variables.

Reviewer 2: Maybe replace `TestBase` with `*TestBase.java`

> Depending on which package you are testing, it may need special environment variables to test succesfully. All packages have a TestBase file inside their corresponding test folder and each one of these contain the special environment variables the tests need in order to run. Make sure to set these variables before running the tests themselves. You may need to restart your development environment after creating or updating these environment variables.

Reviewer 3: Typo: ```succesfully``` should be ```successfully```

### Samples and building

As your code changes, the samples inside the `samples` folder should also change. The README file extracts code from the samples to avoid duplicated code. For a more in-depth look on how to extract the code from the samples into the README file, check this [guide]https://github.com/Azure/azure-sdk-for-java/wiki/Building#code-snippets-in-readme)

That guide also serves as a great starting point for ideas on how to run the code locally to avoid problems on the pipeline whenever you upload your changes. Make sure to give that document a read too.

## Submitting a Pull Request

The easiest way for you to test and not worry about any breaking changes you may cause is to create a fork from the [Java Azure SDK repo](https://github.com/Azure/azure-sdk-for-java). After downloading your repo, make sure to add the original repo as an upstream. To do this, use the `git remote add upstream https://github.com/Azure/azure-sdk-for-java`.

> Create a branch for any new feature you may want to add and when your changes are ready, push your branch to the origin. Because the upstream was already set, if you go to the Python Azure SDK repo you will see a message saying that you pushed changes to your fork and give you the option to create a PR from your fork to the original repo.

Reviewer 4: ```suggestion
Create a branch for any new feature you may want to add and when your changes are ready, push your branch to the origin. Because the upstream was already set, if you go to the Java Azure SDK repo you will see a message saying that you pushed changes to your fork and give you the option to create a PR from your fork to the original repo.
```

Author: Good catch, thank you!
